lua/dap.lua

```diff
@@ -848,24 +863,68 @@ function M.restart()
   end
 end
 
-function M.toggle_breakpoint(condition)
+function M.list_breakpoints(open_quickfix)
+  local bp_signs = get_breakpoint_signs()
+  local num_bufs = #bp_signs
+  local qf_list = {}
+  for _, buf_bp_signs in pairs(bp_signs) do
+    local bufnr = buf_bp_signs.bufnr
+    for _, bp in pairs(buf_bp_signs.signs) do
+      local condition = bp_info[bp.id].condition;
+      local hitCondition = bp_info[bp.id].hitCondition;
+      local logMessage = bp_info[bp.id].logMessage;
+      local text = table.concat({
+               unpack(api.nvim_buf_get_lines(bufnr, bp.lnum - 1, bp.lnum, false), 1),
+               non_empty_sequence(logMessage) and "Log message: "..logMessage,
+               non_empty_sequence(condition) and "Condition: "..condition,
+               non_empty_sequence(hitCondition) and "Hit condition: "..hitCondition,
+             }, ', ')
+      table.insert(qf_list, {
+        bufnr = bufnr,
+        lnum = bp.lnum,
+        col = 0,
+        text = text,
+      })
+    end
+  end
+
+  vim.fn.setqflist({}, 'r', {items = qf_list, context = "DAP Breakpoints", title = "DAP Breakpoints" })
+  if open_quickfix ~= false then
+    if num_bufs == 0 then
+      print('No breakpoints set!')
+    else
+      api.nvim_command('copen')
+    end
+  end
+end
+
+function M.set_breakpoint(condition, hit_condition, log_message)
+  M.toggle_breakpoint(condition, hit_condition, log_message, true)
+end
+
+function M.toggle_breakpoint(condition, hit_condition, log_message, replace_old)
   local bufnr = api.nvim_get_current_buf()
   local lnum, _ = unpack(api.nvim_win_get_cursor(0))
-  if not remove_breakpoint_signs(bufnr, lnum) then
+  if not remove_breakpoint_signs(bufnr, lnum) or replace_old then
     local sign_id = vim.fn.sign_place(
       0,
       ns_breakpoints,
-      'DapBreakpoint',
+      non_empty_sequence(log_message) and 'DapLogPoint' or 'DapBreakpoint',
       bufnr,
       { lnum = lnum; priority = 11; }
     )
-    if condition and sign_id ~= -1 then
-      bp_conditions[sign_id] = condition
+    if sign_id ~= -1 then
+      bp_info[sign_id] = {
+        condition = condition,
+        logMessage = log_message,
+        hitCondition = hit_condition
+      }
     end
   end
   if session and session.initialized then
     session:set_breakpoints(bufnr)
   end
+  M.list_breakpoints(false)
```

Comment:
Not sure if I like `toggle_breakpoint` to automatically update the quickfix list and potentially override other things. I think if a user likes this behavior, it would be easy enough to create a mapping that calls both functions.

Would you be okay with removing this?

Reply:
Sure, I was also not sure how to handle this. But at least we are now overwriting other quick fixes when using our own quick fix context. It could be that we set the focus to our quick fix since it's the newest. We could only do it if our quick fix is already populated.

But I'm also fine with removing. Users can also update it after the setBreakpoints response if they want to.

Reply:
> We could only do it if our quick fix is already populated.

You mean if a user triggered `list_breakpoints` and the current quickfix window contains breakpoints, toggling a breakpoint would update the list. But if the current quickfix window contains anything else it wouldn't update the list?

That sounds also good to me :+1:

Reply:
The changed version will only update the qflist if it's the current one. It will however not update the quickfix list if it's been replaced by a newer quickfix list by another plugin.

```
:help getqflist-examples
:help colder
```

Reply:
Should we auto-close when our quickfix list is empty? I guess it's better to leave this to the user.

Reply:
> Should we auto-close when our quickfix list is empty? I guess it's better to leave this to the user.

Agreed, I'd leave that to the user as well.